autogalaxy/profiles/mass_profiles/calc_lens.py

```diff
@@ -0,0 +1,749 @@
+
+        `convergence = 1.0 - 0.5 * (jacobian_{0,0} + jacobian_{1,1}) = 0.5 * (jacobian_xx + jacobian_yy)`
+
+        By going via the Jacobian, the convergence must be calculated using 2D uniform grid.
+
+        This calculation of the convergence is independent of analytic calculations defined within `MassProfile`
+        objects and the calculation via the Hessian. It can therefore be used as a cross-check.
+
+        Parameters
+        ----------
+        grid
+            The 2D grid of (y,x) arc-second coordinates the deflection angles and Jacobian are computed on.
+        jacobian
+            A precomputed lensing jacobian, which is passed throughout the `CalcLens` functions for efficiency.
+        """
+        convergence = 1 - 0.5 * (jacobian[0][0] + jacobian[1][1])
+
+        return aa.Array2D(array=convergence, mask=grid.mask)
+
+    @precompute_jacobian
+    def shear_yx_2d_via_jacobian_from(self, grid, jacobian=None) -> Union[ShearYX2D, ShearYX2DIrregular]:
+        """
+        Returns the 2D (y,x) shear vectors of the lensing object, which are computed from the 2D deflection angle map
+        via the Jacobian using the expression (see equation 58 https://www.tau.ac.il/~lab3/MICROLENSING/JeruLect.pdf):
+
+        `shear_y = -0.5 * (jacobian_{0,1} + jacobian_{1,0} = -0.5 * (jacobian_yx + jacobian_xy)`
+        `shear_x = 0.5 * (jacobian_{1,1} + jacobian_{0,0} = 0.5 * (jacobian_yy + jacobian_xx)`
+
+        By going via the Jacobian, the convergence must be calculated using 2D uniform grid.
+
+        This calculation of the shear vectors is independent of analytic calculations defined within `MassProfile`
+        objects and the calculation via the Hessian. It can therefore be used as a cross-check.
+
+        Parameters
+        ----------
+        grid
+            The 2D grid of (y,x) arc-second coordinates the deflection angles and Jacobian are computed on.
+        jacobian
+            A precomputed lensing jacobian, which is passed throughout the `CalcLens` functions for efficiency.
+        """
+
+        shear_yx_2d = np.zeros(shape=(grid.sub_shape_slim, 2))
+        shear_yx_2d[:, 0] = -0.5 * (jacobian[0][1] + jacobian[1][0])
+        shear_yx_2d[:, 1] = 0.5 * (jacobian[1][1] - jacobian[0][0])
+
+        if isinstance(grid, aa.Grid2DIrregular):
+            return ShearYX2DIrregular(vectors=shear_yx_2d, grid=grid)
+        return ShearYX2D(vectors=shear_yx_2d, grid=grid, mask=grid.mask)
+
+    def add_functions(self, obj):
+        """
+        This passes all functions of the `_calc_lens` property to the mass object using it.
+
+        This means that instead of having to call a function using the full path:
+
+        `mass_object._calc_lens.magnification_2d_from`
+
+        We can simply call it using the path:
+
+        `mass_object.magnification_2d_from`
+        """
+        obj.tangential_eigen_value_from = self.tangential_eigen_value_from

```

Comment:
This is a bit whacky but hopefully it won't cause an issue.

I think you could do it more concisely:
```python
for name, value in self.__dict__.items():
    if isinstance(value, types.FunctionType):
        setattr(obj, value.__name__, value)